### PR TITLE
[c10d] Fix bfloat16 NCCL PREMUL_SUM factor being read as zero

### LINKED ISSUE #190745: [c10d] bfloat16 NCCL PREMUL_SUM silently zeroes the reduction for power-of-two factors
## Bug description

`ReduceOp.PREMUL_SUM(factor)` over a bfloat16 tensor multiplies by an incorrect factor: for a host (Python float) scalar whose low 2 bytes are zero, which includes every power-of-two value, the reduction comes back all zeros. This is what FSDP2 hits: with a custom `gradient_divide_factor`, `_get_gradient_divide_factors` issues `ReduceOp.PREMUL_SUM(1 / factor)` over bf16 gradients (`torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py`), so a factor of 2, 4, 8, ... produces exact-zero gradients.

## Repro (2 GPUs)

```python
# torchrun --standalone --nproc_per_node=2 repro.py
import torch, torch.distributed as dist, os
dist.init_process_group("nccl")
torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
rank, ws = dist.get_rank(), dist.get_world_size()

for dtype in (torch.float16, torch.float32, torch.bfloat16):
    v = torch.full((8,), rank + 1.0, device="cuda", dtype=dtype)
    dist.all_reduce(v, op=dist.ReduceOp.PREMUL_SUM(0.5))
    expected = 0.5 * sum(i + 1 for i in range(ws))
    print(dtype, "got", v.float().mean().item(), "expected", expected)
```

On 2 ranks (NCCL >= 2.11.1):

```
torch.float16  got 1.5  expected 1.5
torch.float32  got 1.5  expected 1.5
torch.bfloat16 got 0.0  expected 1.5   <-- wrong
```

Verified failing on torch 2.9.1+cu128 and a 2.12 nightly on 2x H100. bf16 device-tensor factors are separately rejected with a dtype error.

## Root cause

`getNcclReduceOp` in `torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp` dispatches the bfloat16 case through `unpackPreMulSum<float, ncclBfloat16>`. `ncclRedOpCreatePreMulSum` reads `ncclTypeSize(dataType)` bytes from the factor pointer, i.e. 2 bytes for `ncclBfloat16`, but the factor is stored as a 4-byte `float scalar_factor`. On little-endian hardware NCCL reads the low 2 bytes; for any float whose low half is zero (all powers of two, e.g. 0.5 = 0x3F000000, low bytes 0x0000) it receives a bfloat16 zero. The `half` and `float` cases are correct because there the scalar type width matches the reduction datatype. The device-factor path fails differently: `const_data_ptr<float>()` requires a float tensor, so a bf16 factor tensor raises.

## Fix

Use `at::BFloat16` as the scalar type for the bfloat16 case (matching the `at::Half` / `float` cases), so the factor is the correct 2-byte width for both the host-immediate and device residences. One line in `getNcclReduceOp`, plus extending the PreMulSum coverage in `test/distributed/test_c10d_ops_nccl.py` (currently `half, float, double`) to include `bfloat16`.

## Reachability

Requires FSDP2 with a custom `gradient_divide_factor` that differs from the reduce-scatter group size, and bf16 `reduce_dtype` (the FSDP2 default takes the AVG path and is safe). This is the expert-parallel / MoE gradient-scaling pattern (e.g. PrimeRL sets `set_gradient_divide_factor` on expert FSDP units). Direct `ReduceOp.PREMUL_SUM(host_scalar)` on any bf16 collective is affected regardless of FSDP2.

## Prior art

Distinct from #155903 / #155904 / #155915 (PreMulSum tensor-factor plumbing). No issue found for the bf16 host-scalar width bug.

*This report was authored with assistance from Claude; the repro and root cause were human-reviewed. Found in a correctness audit alongside a parallel independent audit.*

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh

### PR DESCRIPTION
**Pull request by bogatyy**
Fixes #190745.

## Summary

`getNcclReduceOp` dispatched bfloat16 `PreMulSum` through `unpackPreMulSum<float, ncclBfloat16>`. NCCL reads `ncclTypeSize(dataType)` = 2 bytes from the scalar factor, but it was stored as a 4-byte `float`. For a little-endian float whose low 2 bytes are zero (every power of two, including the `1/factor` FSDP2 passes for a custom `gradient_divide_factor`), NCCL received a bfloat16 zero and silently produced all-zero gradients; a bfloat16 device factor was also rejected because `const_data_ptr<float>` requires a float tensor. Use `at::BFloat16` so the scalar is the correct width for both residences.

## Change

- `torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp`: `unpackPreMulSum<float, ncclBfloat16>` -> `unpackPreMulSum<at::BFloat16, ncclBfloat16>`.
- `test/distributed/test_c10d_ops_nccl.py`: extend the PreMulSum dtype loop in `test_allreduce_ops` from `half, float, double` to include `bfloat16` (host scalar 3.0 and device tensor 5.0). The host-scalar case returns 0 before the fix and the correct value after.

## Verification

CI (the NCCL multi-GPU distributed test jobs); the author has no local CUDA/NCCL build, so this is verification-pending on those jobs. Happy to attach a 2-GPU run before merge.

```
python test/distributed/test_c10d_ops_nccl.py -k test_allreduce_ops
```

*This PR was authored with assistance from Claude; the fix was human-reviewed.*

### CODE REVIEW

**Comment by bogatyy** [comment]
Verified on 2x L40S (NCCL 2.30.7) with a same-hardware A/B, changing only the one line in `getNcclReduceOp`:

| build | fp16 | fp32 | bf16 (host scalar 0.5) | bf16 (device factor) |
|---|---|---|---|---|
| `unpackPreMulSum<float, ncclBfloat16>` (before) | 1.5 | 1.5 | **0.0** | **RuntimeError** |
| `unpackPreMulSum<at::BFloat16, ncclBfloat16>` (this PR) | 1.5 | 1.5 | 1.5 | 1.5 |

`PREMUL_SUM(0.5)` all-reduce over world size 2, expected 1.5 for every dtype. Before the fix the bf16 host scalar reads as zero (low 2 bytes of the 4-byte float) and a bf16 device factor is rejected by `const_data_ptr<float>`; both are correct after. `test_c10d_ops_nccl.py -k test_allreduce_ops` (now including the bfloat16 premul case) passes on the fixed build. This also confirms the bug is still live on current NCCL 2.30.7, not just the older NCCL the report was found on.

**Review by kapilsh — APPROVED** [review]
looks good. thank you.

File: test/distributed/test_c10d_ops_nccl.py [hunk 1]
```diff
@@ -197,7 +197,10 @@ def allreduce(tensors, op):
 
         # Premul Sum
         if torch.cuda.nccl.version() >= (2, 11, 1):
-            for dtype in torch.half, torch.float, torch.double:
+            # bfloat16 exercises the host-scalar path that silently zeroed the
+            # reduction when the factor was passed as a 4-byte float; the
+            # tensor factor covers the device-scalar path.
```
Comment (kapilsh) [inline test/distributed/test_c10d_ops_nccl.py]: nit: claude is over-eager to add extra comments - we can probably remove this since you documented it in PGNCCL.

**Comment by bogatyy** [comment]
@kapilsh I've addressed your comments, PTAL.
(I also may not have notified you correctly about it initially – let me know what's the best practice, say a comment in the conversation or "re-request review")
